### PR TITLE
plan: build separate required physical property for children

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/8634

### What is changed and how it works?

RCA is provided here https://github.com/pingcap/tidb/issues/8634#issuecomment-445703564

For logical operator which passes the required physical property down to its children, make a copy for it instead of direct assignment using the original pointer.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8635)

### CODE REVIEW

**Comment by eurekaka** [comment]
@zz-jason @winoros PTAL

**Comment by eurekaka** [comment]
@lamxTyler @XuHuaiyu PTAL

File: planner/property/physical_property.go [hunk 1]
```diff
@@ -101,3 +101,14 @@ func (p *PhysicalProperty) HashCode() []byte {
...
 	return fmt.Sprintf("Prop{cols: %v, desc: %v, TaskTp: %s, expectedCount: %v}", p.Cols, p.Desc, p.TaskTp, p.ExpectedCnt)
 }
+
+// Clone returns a copy of PhysicalProperty.
+func (p *PhysicalProperty) Clone() *PhysicalProperty {
+	prop := &PhysicalProperty{
+		Cols:        p.Cols,
+		Desc:        p.Desc,
+		TaskTp:      p.TaskTp,
+		ExpectedCnt: p.ExpectedCnt,
```
Comment (lamxTyler) [inline planner/property/physical_property.go]: Add a comment for why not copy `enforced`?

Comment (eurekaka) [reply planner/property/physical_property.go]: Comment addressed, PTAL

**Comment by zz-jason** [comment]
@eurekaka Please cherry pick this change to release-2.1 and release-2.0.

**Comment by eurekaka** [comment]
Sure

**Pull request merged**